src/main/kotlin/org/jitsi/nlj/transform/node/incoming/TccGeneratorNode.kt

```diff
@@ -58,65 +71,150 @@ class TccGeneratorNode(
     }
     private var numTccSent: Int = 0
 
+    init {
+        reschedule()
+    }
+
     override fun observe(packetInfo: PacketInfo) {
         tccExtensionId?.let { tccExtId ->
             val rtpPacket = packetInfo.packetAs<RtpPacket>()
             rtpPacket.getHeaderExtension(tccExtId)?.let { ext ->
                 val tccSeqNum = TccHeaderExtension.getSequenceNumber(ext)
-                addPacket(tccSeqNum, packetInfo.receivedTime, rtpPacket.isMarked)
+                synchronized(lock) {
+                    addPacket(tccSeqNum, packetInfo.receivedTime, rtpPacket.isMarked)
+                }
             }
         }
     }
 
-    override fun handleEvent(event: Event) {
-        when (event) {
-            is RtpExtensionAddedEvent -> {
-                if (event.rtpExtension.type == TRANSPORT_CC) {
-                    tccExtensionId = event.rtpExtension.id.toUInt()
-                    logger.cinfo { "TCC generator setting extension ID to $tccExtensionId" }
-                }
+    //TODO: we don't handle tcc seq num rollover here, but the old code didn't seem to either?
+    private fun addPacket(tccSeqNum: Int, timestamp: Long, isMarked: Boolean) {
+        if (packetArrivalTimes.ceilingKey(windowStartSeq) == null) {
+            // Packets in map are all older than the start of the next tcc feedback packet,
+            // remove them
+            //TODO: chrome does something more advanced. is this good enough?
+            packetArrivalTimes.clear()
+        }
+        if (windowStartSeq == -1) {
+            windowStartSeq = tccSeqNum
+        } else if (tccSeqNum < windowStartSeq) {
+            windowStartSeq = tccSeqNum
+        }
+        packetArrivalTimes.putIfAbsent(tccSeqNum, timestamp)
+        if (!periodicFeedbacks && isTccReadyToSend(isMarked)) {
+            buildAndSendFeedback()
+        }
+    }
+
+    private fun sendPeriodicFeedbacks() {
+        try {
+            synchronized(lock) {
+                logger.cdebug { "${System.identityHashCode(this)} sending periodic feedback at " +
+                        "${ java.lang.System.currentTimeMillis()}, window start seq is $windowStartSeq" }
+                buildAndSendFeedback()
             }
-            is RtpExtensionClearEvent -> tccExtensionId = null
-            is ReceiveSsrcAddedEvent -> mediaSsrcs.add(event.ssrc)
-            is ReceiveSsrcRemovedEvent -> mediaSsrcs.remove(event.ssrc)
+        } catch (t: Throwable) {
+            logger.error("Error sending feedback", t)
+        } finally {
+            reschedule()
         }
     }
 
-    private fun addPacket(tccSeqNum: Int, timestamp: Long, isMarked: Boolean) {
-        if (currTccBuilder.addPacket(tccSeqNum, timestamp)) {
-            if (isTccReadyToSend(isMarked)) {
-                sendTcc()
+    private fun buildAndSendFeedback() {
```

Comment:
This executes while holding the lock. Can we split it so that `sendTcc` (which calls the external callback) doesn't hold the lock?